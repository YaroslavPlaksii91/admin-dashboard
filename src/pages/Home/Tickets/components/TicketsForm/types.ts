import { TicketType } from '../TicketsItem/types';

export type TicketsFormData = {
  title: string;
  name: string;
  date: {
    $d: string;
  };
  priority: string;
};

export type TicketsFormProps = {
  addTicket: (data: TicketsFormData) => void;
  editTicket: (data: TicketsFormData) => void;
  activeTicket?: TicketType;
};
