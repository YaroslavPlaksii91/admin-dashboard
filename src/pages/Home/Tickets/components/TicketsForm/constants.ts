export const TICKETS_FORM_FIELDS = Object.freeze({
  TITLE: 'title',
  NAME: 'name',
  DATE: 'date',
  PRIORITY: 'priority',
});

export const TICKETS_FORM_CONFIG = Object.freeze({
  [TICKETS_FORM_FIELDS.TITLE]: {
    label: 'Ticket details',
    type: 'text',
    name: 'title',
    placeholder: 'Add description',
  },
  [TICKETS_FORM_FIELDS.NAME]: {
    label: 'Customer name',
    type: 'text',
    name: 'name',
    placeholder: 'Name',
  },
  [TICKETS_FORM_FIELDS.DATE]: {
    label: 'Date',
    name: 'date',
  },
  [TICKETS_FORM_FIELDS.PRIORITY]: {
    label: 'Priority',
    name: 'priority',
    options: [
      {
        value: 'priority',
        label: 'Choose priority',
        disabled: true,
      },
      {
        value: 'high',
        label: 'High',
      },
      {
        value: 'low',
        label: 'Low',
      },
      {
        value: 'normal',
        label: 'Normal',
      },
    ],
  },
});
