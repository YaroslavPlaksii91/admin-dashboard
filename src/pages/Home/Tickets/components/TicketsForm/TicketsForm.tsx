import { FC } from 'react';
import { useForm, SubmitHandler, Controller } from 'react-hook-form';
import { Button, Box } from '@mui/material';

import { FormInput } from '@components/FormInput/FormInput';
import { Select } from '@components/Select/Select';
import { DatePicker } from '@components/DatePicker/DatePicker';
import { PRIORITY_OPTIONS } from '@pages/Home/Tickets/constants';

import { TICKETS_FORM_CONFIG, TICKETS_FORM_FIELDS } from './constants';
import { TicketsFormData, TicketsFormProps } from './types';

export const TicketsForm: FC<TicketsFormProps> = ({
  addTicket,
  editTicket,
  activeTicket,
}) => {
  const {
    register,
    handleSubmit,
    control,
    formState: { errors },
  } = useForm<TicketsFormData>();

  const onSubmit: SubmitHandler<TicketsFormData> = data => {
    if (activeTicket) {
      editTicket({ ...data, date: { $d: activeTicket.date } });
    } else {
      addTicket(data);
    }
  };

  return (
    <Box
      component="form"
      onSubmit={handleSubmit(onSubmit)}
      noValidate
      sx={{ margin: '70px 0 24px' }}
    >
      <FormInput
        {...TICKETS_FORM_CONFIG[TICKETS_FORM_FIELDS.TITLE]}
        register={register('title', {
          required: 'Title is required',
          minLength: {
            value: 5,
            message: 'Title must be at least 5 characters',
          },
        })}
        errors={errors}
        defaultValue={activeTicket?.title}
      />

      <FormInput
        {...TICKETS_FORM_CONFIG[TICKETS_FORM_FIELDS.NAME]}
        register={register('name', {
          required: 'Name is required',
          minLength: {
            value: 2,
            message: 'Name must be at least 2 characters',
          },
        })}
        errors={errors}
        defaultValue={activeTicket?.customerName}
      />

      <Controller
        control={control}
        name={TICKETS_FORM_FIELDS.DATE}
        rules={{
          required: {
            value: true && !Boolean(activeTicket),
            message: 'Date is required',
          },
        }}
        render={({ field }) => {
          return (
            <DatePicker
              field={field}
              errors={errors}
              disabled={Boolean(activeTicket)}
              {...TICKETS_FORM_CONFIG[TICKETS_FORM_FIELDS.DATE]}
            />
          );
        }}
      />

      <Select
        {...TICKETS_FORM_CONFIG[TICKETS_FORM_FIELDS.PRIORITY]}
        register={register('priority', {
          validate: value =>
            PRIORITY_OPTIONS.includes(value) || 'Choose priority',
        })}
        errors={errors}
        defaultValue={activeTicket?.priority}
      />

      <Button variant="contained" type="submit" fullWidth>
        Save
      </Button>
    </Box>
  );
};
