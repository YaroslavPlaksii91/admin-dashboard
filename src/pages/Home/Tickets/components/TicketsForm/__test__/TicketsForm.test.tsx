import {
  cleanup,
  render,
  screen,
  fireEvent,
  waitFor,
} from '@testing-library/react';

import { TicketsForm } from '../TicketsForm';

describe('TicketsForm component', () => {
  afterEach(cleanup);

  const addTicketMock = jest.fn();
  const editTicketMock = jest.fn();

  it('should be rendered', () => {
    const { asFragment } = render(
      <TicketsForm addTicket={addTicketMock} editTicket={editTicketMock} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders form fields correctly', () => {
    render(
      <TicketsForm addTicket={addTicketMock} editTicket={editTicketMock} />,
    );

    const titleField = screen.getByLabelText('Ticket details');
    const nameField = screen.getByLabelText('Customer name');
    const dateField = screen.getByLabelText('Date');
    const priorityField = screen.getByLabelText('Priority');

    expect(titleField).toBeInTheDocument();
    expect(nameField).toBeInTheDocument();
    expect(dateField).toBeInTheDocument();
    expect(priorityField).toBeInTheDocument();
  });

  it('displays validation error messages', async () => {
    render(
      <TicketsForm addTicket={addTicketMock} editTicket={editTicketMock} />,
    );

    const titleField = screen.getByLabelText('Ticket details');
    const nameField = screen.getByLabelText('Customer name');
    const submitBtn = screen.getByRole('button', { name: 'Save' });

    fireEvent.change(titleField, { target: { value: '' } });
    fireEvent.change(nameField, { target: { value: 'A' } });

    fireEvent.click(submitBtn);

    await waitFor(() => {
      const titleError = screen.getByText('Title is required');
      const nameError = screen.getByText('Name must be at least 2 characters');

      expect(titleError).toBeInTheDocument();
      expect(nameError).toBeInTheDocument();
    });
  });
});
