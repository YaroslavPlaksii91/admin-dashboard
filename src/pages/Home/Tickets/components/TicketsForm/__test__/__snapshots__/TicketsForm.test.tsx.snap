// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketsForm component should be rendered 1`] = `
<DocumentFragment>
  <form
    class="MuiBox-root css-m9nphe"
    novalidate=""
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r0:"
        id=":r0:-label"
      >
        Ticket details
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          name="title"
          placeholder="Add description"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Ticket details
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r1:"
        id=":r1:-label"
      >
        Customer name
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r1:"
          name="name"
          placeholder="Name"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Customer name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r2:"
        id=":r2:-label"
      >
        Date
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r2:"
          inputmode="text"
          placeholder="MM/DD/YYYY hh:mm aa"
          type="text"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <button
            aria-label="Choose date"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CalendarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Date
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r4:"
        id=":r4:-label"
      >
        Priority
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby=":r4:-label :r4:"
          class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r4:"
          role="button"
          tabindex="0"
        >
          Choose priority
        </div>
        <input
          aria-hidden="true"
          aria-invalid="false"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          name="priority"
          tabindex="-1"
          value="priority"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Priority
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="submit"
    >
      Save
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
